src/cachemachine/rubintag.py

```diff
@@ -0,0 +1,598 @@
+                    build=build,
+                )
+            except TypeError as exc:
+                logger.warning(f"Could not make semver from tag {tag}: {exc}")
+            name = f"{typename} {restname}"  # Glue together display name.
+            if cycle:
+                name += f" (SAL Cycle {cycle}, Build {cbuild})"
+            if rest:
+                name += f" [{rest}]"
+            cycle_int = RubinPartialTag.maybe_int(cycle)
+        return (name, semver, cycle_int)
+
+    @staticmethod
+    def maybe_int(n: Optional[str]) -> Optional[int]:
+        if n is None:
+            return None
+        return int(n)
+
+    @staticmethod
+    def trailing_parts_to_semver_build_component(
+        cycle: Optional[str],
+        cbuild: Optional[str],
+        ctag: Optional[str],  # if present, either 'c' or 'csal'
+        rest: Optional[str] = None,
+    ) -> Optional[str]:
+        """This takes care of massaging the cycle components, and 'rest', into
+        a semver-compatible buildstring, which is dot-separated and can only
+        contain alphanumerics.  See SQR-059 for how it's used.
+        (https://github.com/lsst-sqre/sqr-059)
+        """
+        if cycle:
+            if rest:
+                # Cycle must always precede rest
+                rest = f"{ctag}{cycle}.{cbuild}_{rest}"
+            else:
+                rest = f"{ctag}{cycle}.{cbuild}"
+        # We're done with cycle components now.
+        if not rest:
+            return None
+        rest = rest.replace("_", ".")
+        pat = re.compile(r"[^\w|\.]+")  # Identify all non alphanum, non-dots
+        # Throw away all of those after turning underscores to dots.
+        rest = pat.sub("", rest)
+        if not rest:  # if we are left with an empty string, return None
+            return None
+        return rest
+
+    def compare(self, other: "RubinPartialTag") -> int:
+        """This is modelled after semver.compare, but raises an exception
+        if the images do not have the same image_type."""
+        if self.image_type != other.image_type:
+            raise IncomparableImageTypesError(
+                f"Tag '{self.tag}' of type {self.image_type} cannot be "
+                + f"compared to '{other.tag}' of type {other.image_type}."
+            )
+        # The easy case: we have a type with a semantic_version attribute.
+        # Use it.
+        if (
+            self.semantic_version is not None
+            and other.semantic_version is not None
+        ):
+            return self.semantic_version.compare(other.semantic_version)
+        # Otherwise, all we can do is sort lexigraphically by tag.
+        # Experimentals can be sorted only by tag.
+        if self.tag == other.tag:
+            return 0
+        if self.tag < other.tag:
+            return -1
+        return 1
+        # We should not be able to get here, but the typechecker can't
+        # prove that.
+        raise IncomparableImageTypesError(
```

Comment:
seems like the return 1 above should handle that? I'm fine with either of these but both seems a bit confusing

Reply:
Yep, left over from some earlier iteration.